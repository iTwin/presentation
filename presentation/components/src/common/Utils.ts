/*---------------------------------------------------------------------------------------------
* Copyright (c) 2019 Bentley Systems, Incorporated. All rights reserved.
* Licensed under the MIT License. See LICENSE.md in the project root for license terms.
*--------------------------------------------------------------------------------------------*/
/** @module Core */

import * as React from "react";
import { I18NNamespace } from "@bentley/imodeljs-i18n";
import { Presentation } from "@bentley/presentation-frontend";

/**
 * An interface of something that has a priority.
 * @internal
 */
export interface IPrioritized {
  priority: number;
}

/**
 * A sorting algorithm for `Array.sort` that sorts items by priority.
 * Higher priority items appear first in the list.
 *
 * @internal
 */
export const prioritySortFunction = (a: IPrioritized, b: IPrioritized): number => {
  if (a.priority > b.priority)
    return -1;
  if (a.priority < b.priority)
    return 1;
  return 0;
};

let localizationNamespace: I18NNamespace | undefined;
/**
 * Translate a string with the specified id from `PresentationComponents`
 * localization namespace. The `stringId` should not contain namespace - it's
 * prepended automatically.
 *
 * @internal
 */
export const translate = async (stringId: string): Promise<string> => {
  const localizationNamespaceName = "PresentationComponents";
  if (!localizationNamespace) {
    localizationNamespace = Presentation.i18n.registerNamespace(localizationNamespaceName);
  }
  await localizationNamespace.readFinished;
  stringId = `${localizationNamespaceName}:${stringId}`;
  return Presentation.i18n.translate(stringId);
};

/**
 * Creates a display name for the supplied component
 * @internal
 */
export const getDisplayName = <P>(component: React.ComponentType<P>): string => {
  if (component.displayName)
    return component.displayName;
  if (component.name)
    return component.name;
  return "Component";
};
