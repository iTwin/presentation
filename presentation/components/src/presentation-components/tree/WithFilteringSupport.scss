/*---------------------------------------------------------------------------------------------
* Copyright (c) Bentley Systems, Incorporated. All rights reserved.
* See LICENSE.md in the project root for license terms and full copyright notice.
*--------------------------------------------------------------------------------------------*/

.filteredTree {

  position: relative;
  width: 100%;
  height: 100%;

  .filteredTreeOverlay {
    position: absolute;
    top: 0;
    left: 0;
    width: 100%;
    height: 100%;
    z-index: 9999;
    background-color: rgba(255,255,255,.8);
  }

}
