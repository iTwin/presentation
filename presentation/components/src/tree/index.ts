/*---------------------------------------------------------------------------------------------
* Copyright (c) 2018 Bentley Systems, Incorporated. All rights reserved.
* Licensed under the MIT License. See LICENSE.md in the project root for license terms.
*--------------------------------------------------------------------------------------------*/
/** @module Tree */

export { default as PresentationTreeDataProvider } from "./DataProvider";
export { default as IPresentationTreeDataProvider } from "./IPresentationTreeDataProvider";
export { default as withUnifiedSelection, SelectionTarget } from "./WithUnifiedSelection";
export { default as withFilteringSupport } from "./WithFilteringSupport";
