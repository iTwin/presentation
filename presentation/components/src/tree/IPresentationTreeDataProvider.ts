/*---------------------------------------------------------------------------------------------
* Copyright (c) 2019 Bentley Systems, Incorporated. All rights reserved.
* Licensed under the MIT License. See LICENSE.md in the project root for license terms.
*--------------------------------------------------------------------------------------------*/
/** @module Tree */

import { ITreeDataProvider, TreeNodeItem } from "@bentley/ui-components";
import { IPresentationDataProvider } from "../common/IPresentationDataProvider";
import { NodeKey, NodePathElement } from "@bentley/presentation-common";

/**
 * Presentation tree data provider.
 * @public
 */
export interface IPresentationTreeDataProvider extends ITreeDataProvider, IPresentationDataProvider {
  /**
   * Returns a [[NodeKey]] from given [[TreeNodeItem]].
   */
  getNodeKey(node: TreeNodeItem): NodeKey;

  /**
   * Returns filtered node paths.
   */
  getFilteredNodePaths(filter: string): Promise<NodePathElement[]>;

  /**
   * Loads the hierarchy so on-demand requests and filtering works quicker
   * @alpha
   */
  loadHierarchy?(): Promise<void>;
}
