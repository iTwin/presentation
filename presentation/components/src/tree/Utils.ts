/*---------------------------------------------------------------------------------------------
* Copyright (c) Bentley Systems, Incorporated. All rights reserved.
* See LICENSE.md in the project root for license terms and full copyright notice.
*--------------------------------------------------------------------------------------------*/
/** @packageDocumentation
 * @module Tree
 */

import { StyleHelper } from "../common/StyleHelper";
import { CheckBoxState } from "@bentley/ui-core";
import { Node, PageOptions as PresentationPageOptions, LabelDefinition } from "@bentley/presentation-common";
import { DelayLoadedTreeNodeItem, PageOptions as UiPageOptions, ItemStyle, ItemColorOverrides } from "@bentley/ui-components";
import { createLabelRecord } from "../common/Utils";

/** @internal */
export const PRESENTATION_TREE_NODE_KEY = "__presentation-components/key";

/** @internal */
export const createTreeNodeItems = (nodes: ReadonlyArray<Readonly<Node>>, parentId?: string): DelayLoadedTreeNodeItem[] => {
  const list = new Array<DelayLoadedTreeNodeItem>();
  for (const node of nodes)
    list.push(createTreeNodeItem(node, parentId));
  return list;
};

/** @internal */
export const createTreeNodeItem = (node: Readonly<Node>, parentId?: string): DelayLoadedTreeNodeItem => {
  const item: DelayLoadedTreeNodeItem = {
    id: [...node.key.pathFromRoot].reverse().join("/"),
    label: node.label,
    labelDefinition: createLabelRecord(node.labelDefinition ? node.labelDefinition : LabelDefinition.fromLabelString(node.label), "node_label"),
  };
  (item as any)[PRESENTATION_TREE_NODE_KEY] = node.key;

  const style: ItemStyle = {};
  const colorOverrides: ItemColorOverrides = {};
  if (parentId)
    item.parentId = parentId;
  if (node.description)
    item.description = node.description;
  if (node.hasChildren)
    item.hasChildren = true;
  if (node.isExpanded)
    item.autoExpand = true;
  if (node.imageId)
    item.icon = node.imageId;
  if (StyleHelper.isBold(node))
    style.isBold = true;
  if (StyleHelper.isItalic(node))
    style.isItalic = true;
  const foreColor = StyleHelper.getForeColor(node);
  if (foreColor)
    colorOverrides.color = foreColor;
  const backColor = StyleHelper.getBackColor(node);
  if (backColor)
    colorOverrides.backgroundColor = backColor;
  if (node.isCheckboxVisible) {
    item.isCheckboxVisible = true;
    if (node.isChecked)
      item.checkBoxState = CheckBoxState.On;
    if (!node.isCheckboxEnabled)
      item.isCheckboxDisabled = true;
  }
  if (Object.keys(colorOverrides).length > 0)
    style.colorOverrides = colorOverrides;
  if (Object.keys(style).length > 0)
    item.style = style;
  if (node.extendedData)
    item.extendedData = node.extendedData;

  return item;
};

/** @internal */
export const pageOptionsUiToPresentation = (pageOptions?: UiPageOptions): PresentationPageOptions | undefined => {
  if (pageOptions)
    return { ...pageOptions };
  return undefined;
};
