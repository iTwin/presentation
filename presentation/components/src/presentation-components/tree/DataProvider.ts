/*---------------------------------------------------------------------------------------------
* Copyright (c) Bentley Systems, Incorporated. All rights reserved.
* See LICENSE.md in the project root for license terms and full copyright notice.
*--------------------------------------------------------------------------------------------*/
/** @packageDocumentation
 * @module Tree
 */

import memoize from "micro-memoize";
import { IDisposable, Logger } from "@bentley/bentleyjs-core";
import { IModelConnection } from "@bentley/imodeljs-frontend";
import { HierarchyRequestOptions, NodeKey, NodePathElement, Ruleset } from "@bentley/presentation-common";
import { Presentation } from "@bentley/presentation-frontend";
import { DelayLoadedTreeNodeItem, PageOptions, TreeNodeItem } from "@bentley/ui-components";
import { RulesetRegistrationHelper } from "../common/RulesetRegistrationHelper";
import { IPresentationTreeDataProvider } from "./IPresentationTreeDataProvider";
import { createTreeNodeItems, pageOptionsUiToPresentation, PRESENTATION_TREE_NODE_KEY } from "./Utils";

/**
 * Properties for creating a `PresentationTreeDataProvider` instance.
 * @public
 */
export interface PresentationTreeDataProviderProps {
  /** IModel to pull data from. */
  imodel: IModelConnection;
  /** Id of the ruleset to use when requesting content or a ruleset itself. */
  ruleset: string | Ruleset;
  /**
   * Paging size for obtaining nodes.
   *
   * Presentation data providers, when used with paging, have ability to save one backend request for size / count. That
   * can only be achieved when `pagingSize` property is set on the data provider and it's value matches size which is used when
   * requesting nodes. To help developers notice this problem, data provider emits a warning similar to this:
   * ```
   * PresentationTreeDataProvider.pagingSize doesn't match pageOptions in PresentationTreeDataProvider.getNodes call. Make sure you set PresentationTreeDataProvider.pagingSize to avoid excessive backend requests.
   * ```
   * To fix the issue, developers should make sure the page size used for requesting data is also set for the data provider:
   * ```TS
   * const pagingSize = 10;
   * const provider = new TreeDataProvider({imodel, ruleset, pagingSize});
   * // only one backend request is made for the two following requests:
   * provider.getNodesCount();
   * provider.getNodes({ start: 0, size: pagingSize });
   * ```
   */
  pagingSize?: number;
}

/**
 * Presentation Rules-driven tree data provider.
 * @public
 */
export class PresentationTreeDataProvider implements IPresentationTreeDataProvider, IDisposable {
  private _imodel: IModelConnection;
  private _rulesetRegistration: RulesetRegistrationHelper;
  private _pagingSize?: number;

  /** Constructor. */
  public constructor(props: PresentationTreeDataProviderProps) {
    this._rulesetRegistration = new RulesetRegistrationHelper(props.ruleset);
    this._imodel = props.imodel;
    this._pagingSize = props.pagingSize;
  }

  /** Destructor. Must be called to clean up.  */
  public dispose() {
    this._rulesetRegistration.dispose();
  }

  /** Id of the ruleset used by this data provider */
  public get rulesetId(): string { return this._rulesetRegistration.rulesetId; }

  /** [[IModelConnection]] used by this data provider */
  public get imodel(): IModelConnection { return this._imodel; }

  /**
   * Paging options for obtaining nodes.
   * @see `PresentationTreeDataProviderProps.pagingSize`
   */
  public get pagingSize(): number | undefined { return this._pagingSize; }
  public set pagingSize(value: number | undefined) { this._pagingSize = value; }

  /** Called to get extended options for node requests */
  private createRequestOptions(): HierarchyRequestOptions<IModelConnection> {
    return {
      imodel: this._imodel,
      rulesetOrId: this._rulesetRegistration.rulesetId,
    };
  }

  /**
   * Returns a [[NodeKey]] from given [[TreeNodeItem]].
   * **Warning:** the `node` must be created by this data provider.
   */
  public getNodeKey(node: TreeNodeItem): NodeKey {
    return (node as any)[PRESENTATION_TREE_NODE_KEY];
  }

  /**
   * Returns nodes
   * @param parentNode The parent node to return children for.
   * @param pageOptions Information about the requested page of data.
   */
  public async getNodes(parentNode?: TreeNodeItem, pageOptions?: PageOptions): Promise<DelayLoadedTreeNodeItem[]> {
    if (undefined !== pageOptions && pageOptions.size !== this.pagingSize) {
      const msg = `PresentationTreeDataProvider.pagingSize doesn't match pageOptions in PresentationTreeDataProvider.getNodes call.
        Make sure you set PresentationTreeDataProvider.pagingSize to avoid excessive backend requests.`;
      Logger.logWarning("Presentation.Components", msg);
    }

    return (await this._getNodesAndCount(parentNode, pageOptions)).nodes;
  }

  /**
   * Returns the total number of nodes
   * @param parentNode The parent node to return children count for.
   */
  public async getNodesCount(parentNode?: TreeNodeItem): Promise<number> {
    const pageOptions = undefined !== this.pagingSize ? { start: 0, size: this.pagingSize } : undefined;
    return (await this._getNodesAndCount(parentNode, pageOptions)).count;
  }

  private _getNodesAndCount = memoize(async (parentNode?: TreeNodeItem, pageOptions?: PageOptions): Promise<{ nodes: TreeNodeItem[], count: number }> => {
    const requestCount = undefined !== pageOptions && 0 === pageOptions.start && undefined !== pageOptions.size;
    const parentKey = parentNode ? this.getNodeKey(parentNode) : undefined;

    if (!requestCount) {
      const allNodes = await Presentation.presentation.getNodes({ ...this.createRequestOptions(), paging: pageOptionsUiToPresentation(pageOptions) }, parentKey);
      return { nodes: parentNode ? createTreeNodeItems(allNodes, parentNode.id) : createTreeNodeItems(allNodes), count: allNodes.length };
    }

    const nodesResponse = await Presentation.presentation.getNodesAndCount({ ...this.createRequestOptions(), paging: pageOptionsUiToPresentation(pageOptions) }, parentKey);
    return { nodes: parentNode ? createTreeNodeItems(nodesResponse.nodes, parentNode.id) : createTreeNodeItems(nodesResponse.nodes), count: nodesResponse.count };
  }, { isMatchingKey: MemoizationHelpers.areNodesRequestsEqual as any });

  /**
   * Returns filtered node paths.
   * @param filter Filter.
   */
  public getFilteredNodePaths = async (filter: string): Promise<NodePathElement[]> => {
    return Presentation.presentation.getFilteredNodePaths(this.createRequestOptions(), filter);
  }

  /**
   * Loads the hierarchy so on-demand requests and filtering works quicker
   * @alpha Hierarchy loading performance needs to be improved before this becomes publicly available.
   */
  public async loadHierarchy() {
    return Presentation.presentation.loadHierarchy(this.createRequestOptions());
  }

}

class MemoizationHelpers {
  public static areNodesRequestsEqual(lhsArgs: [TreeNodeItem?, PageOptions?], rhsArgs: [TreeNodeItem?, PageOptions?]): boolean {
    if (lhsArgs[0]?.id !== rhsArgs[0]?.id)
      return false;
    if ((lhsArgs[1]?.start ?? 0) !== (rhsArgs[1]?.start ?? 0))
      return false;
    if ((lhsArgs[1]?.size ?? 0) !== (rhsArgs[1]?.size ?? 0))
      return false;
    return true;
  }
}
