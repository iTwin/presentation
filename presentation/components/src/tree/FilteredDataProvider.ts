/*---------------------------------------------------------------------------------------------
* Copyright (c) 2019 Bentley Systems, Incorporated. All rights reserved.
* Licensed under the MIT License. See LICENSE.md in the project root for license terms.
*--------------------------------------------------------------------------------------------*/
/** @module Tree */

import { IModelConnection } from "@bentley/imodeljs-frontend";
import { NodePathElement, NodeKey } from "@bentley/presentation-common";
import {
  SimpleTreeDataProvider, SimpleTreeDataProviderHierarchy,
  DelayLoadedTreeNodeItem, TreeNodeItem,
  PageOptions, ActiveMatchInfo,
} from "@bentley/ui-components";
import { createTreeNodeItem } from "./Utils";
import { IPresentationTreeDataProvider } from "./IPresentationTreeDataProvider";
import { memoize } from "lodash";

/**
 * Rules-driven presentation tree data provider that returns filtered results.
 * @internal
 */
export class FilteredPresentationTreeDataProvider implements IPresentationTreeDataProvider {
  private _parentDataProvider: IPresentationTreeDataProvider;
  private _filteredDataProvider: SimpleTreeDataProvider;
  private _filter: string;
  private _filteredResultMatches: Array<{ id: string, matchesCount: number }> = [];

  public constructor(parentDataProvider: IPresentationTreeDataProvider, filter: string, paths: ReadonlyArray<Readonly<NodePathElement>>) {
    this._parentDataProvider = parentDataProvider;
    this._filter = filter;
    const hierarchy: SimpleTreeDataProviderHierarchy = new Map<string | undefined, TreeNodeItem[]>();
    this.createHierarchy(paths, hierarchy);
    this._filteredDataProvider = new SimpleTreeDataProvider(hierarchy);
  }

  public get rulesetId(): string { return this._parentDataProvider.rulesetId; }

  public get imodel(): IModelConnection { return this._parentDataProvider.imodel; }

  public get filter(): string { return this._filter; }

  public get parentDataProvider(): IPresentationTreeDataProvider { return this._parentDataProvider; }

  private createHierarchy(paths: ReadonlyArray<Readonly<NodePathElement>>, hierarchy: SimpleTreeDataProviderHierarchy, parentId?: string) {
    const treeNodes: DelayLoadedTreeNodeItem[] = [];
    for (let i = 0; i < paths.length; i++) {
      const node = createTreeNodeItem(paths[i].node, parentId);

      if (paths[i].filteringData && paths[i].filteringData!.matchesCount)
        this._filteredResultMatches.push({ id: node.id, matchesCount: paths[i].filteringData!.matchesCount });

      if (paths[i].children.length !== 0) {
        this.createHierarchy(paths[i].children, hierarchy, node.id);
        node.hasChildren = true;
        node.autoExpand = true;
      } else {
        delete node.hasChildren;
        delete node.autoExpand;
      }

      treeNodes[i] = node;
    }
    hierarchy.set(parentId, treeNodes);
  }

  public getActiveMatch: (index: number) => ActiveMatchInfo | undefined = memoize((index: number): ActiveMatchInfo | undefined => {
    let activeMatch: ActiveMatchInfo | undefined;
    if (index <= 0)
      return undefined;

    let i = 1;
    for (const node of this._filteredResultMatches) {
      if (index < i + node.matchesCount) {
        activeMatch = {
          nodeId: node.id,
          matchIndex: index - i,
        };
        break;
      }

      i += node.matchesCount;
    }
    return activeMatch;
  });

  /** Count filtering results. Including multiple possible matches within node labels */
  public countFilteringResults(nodePaths: ReadonlyArray<Readonly<NodePathElement>>): number {
    let resultCount = 0;

    // Loops through root level only
    for (const path of nodePaths) {
      if (path.filteringData)
        resultCount += path.filteringData.matchesCount + path.filteringData.childMatchesCount;
    }

    return resultCount;
  }

  public async getNodes(parent?: TreeNodeItem, pageOptions?: PageOptions): Promise<DelayLoadedTreeNodeItem[]> {
    return this._filteredDataProvider.getNodes(parent, pageOptions);
  }

  public async getNodesCount(parent?: TreeNodeItem): Promise<number> {
    return this._filteredDataProvider.getNodesCount(parent);
  }

  public async getFilteredNodePaths(filter: string): Promise<NodePathElement[]> {
    return this._parentDataProvider.getFilteredNodePaths(filter);
  }

  public getNodeKey(node: TreeNodeItem): NodeKey {
    return this._parentDataProvider.getNodeKey(node);
  }

  /** @alpha */
  // istanbul ignore next
  public async loadHierarchy() {
    // the hierarchy is already loaded when this provider is created
  }

}
