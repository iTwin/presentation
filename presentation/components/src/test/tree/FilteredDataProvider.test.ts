/*---------------------------------------------------------------------------------------------
* Copyright (c) Bentley Systems, Incorporated. All rights reserved.
* See LICENSE.md in the project root for license terms and full copyright notice.
*--------------------------------------------------------------------------------------------*/
/* tslint:disable:no-direct-imports */

import { expect } from "chai";
import * as moq from "@bentley/presentation-common/lib/test/_helpers/Mocks";
import * as faker from "faker";
import { createRandomTreeNodeItem } from "../_helpers/UiComponents";
import { createRandomECInstanceNodeKey, createRandomNodePathElement } from "@bentley/presentation-common/lib/test/_helpers/random";
import { NodePathElement } from "@bentley/presentation-common";
import { PageOptions } from "@bentley/ui-components";
import { IModelConnection } from "@bentley/imodeljs-frontend";
import { FilteredPresentationTreeDataProvider } from "../../tree/FilteredDataProvider";
import { IPresentationTreeDataProvider } from "../../tree/IPresentationTreeDataProvider";
import { createTreeNodeItem } from "../../tree/Utils";

describe("FilteredTreeDataProvider", () => {

  /*
  A-1
    A-1-1
  A-2
    A-2-1
    A-2-2
      A-2-2-1
  */
  const nodePaths: NodePathElement[] = [];

  nodePaths[0] = createRandomNodePathElement();
  nodePaths[0].node.label = "A-1";

  nodePaths[1] = createRandomNodePathElement();
  nodePaths[1].node.label = "A-2";

  nodePaths[0].children = [];
  nodePaths[0].children[0] = createRandomNodePathElement();
  nodePaths[0].children[0].node.label = "A-1-1";

  nodePaths[1].children = [];
  nodePaths[1].children[0] = createRandomNodePathElement();
  nodePaths[1].children[0].node.label = "A-2-1";

  nodePaths[1].children[1] = createRandomNodePathElement();
  nodePaths[1].children[1].node.label = "A-2-2";

  nodePaths[1].children[1].children = [];
  nodePaths[1].children[1].children[0] = createRandomNodePathElement();
  nodePaths[1].children[1].children[0].node.label = "A-2-2-1";

  let provider: FilteredPresentationTreeDataProvider;
  let filter: string;
  const parentProviderMock = moq.Mock.ofType<IPresentationTreeDataProvider>();
  const imodelMock = moq.Mock.ofType<IModelConnection>();
  const pageOptions: PageOptions = { size: 0, start: 0 };

  beforeEach(() => {
    parentProviderMock.reset();
    filter = faker.random.word();
    provider = new FilteredPresentationTreeDataProvider(parentProviderMock.object, filter, nodePaths);
  });

  describe("filter", () => {

    it("returns filter with which it was initialized", () => {
      expect(provider.filter).to.be.equal(filter);
    });

  });

  describe("rulesetId", () => {

    it("returns rulesetId of the parent data provider", () => {
      const expectedRulesetId = faker.random.word();
      parentProviderMock.setup((x) => x.rulesetId)
        .returns(() => expectedRulesetId)
        .verifiable();
      expect(provider.rulesetId).to.eq(expectedRulesetId);
      parentProviderMock.verifyAll();
    });

  });

  describe("imodel", () => {

    it("returns imodel of the parent data provider", () => {
      parentProviderMock.setup((x) => x.imodel)
        .returns(() => imodelMock.object)
        .verifiable();
      expect(provider.imodel).to.eq(imodelMock.object);
      parentProviderMock.verifyAll();
    });

  });

  describe("parentDataProvider", () => {

    it("returns parent data provider", () => {
      expect(provider.parentDataProvider).to.eq(parentProviderMock.object);
    });

  });

  describe("getNodes", () => {

    it("returns root nodes", async () => {
      const result = await provider.getNodes(undefined, pageOptions);
      expect(result).to.matchSnapshot();
    });

    it("returns child nodes", async () => {
      const parentNode = createTreeNodeItem(nodePaths[1].node);

      const result = await provider.getNodes(parentNode, pageOptions);
      expect(result).to.matchSnapshot();
    });

  });

  describe("getNodesCount", () => {

    it("returns root nodes count", async () => {
      const result = await provider.getNodesCount();
      expect(result).to.equal(nodePaths.length);
    });

    it("returns child nodes count", async () => {
      const parentNode = createTreeNodeItem(nodePaths[1].node);

      const result = await provider.getNodesCount(parentNode);
      expect(result).to.equal(nodePaths[1].children.length);
    });

  });

  describe("getFilteredNodePaths", () => {

    it("calls parent data provider", async () => {
      parentProviderMock.setup((x) => x.getFilteredNodePaths(filter))
        .returns(async () => nodePaths)
        .verifiable();

      const result = await provider.getFilteredNodePaths(filter);
      expect(result).to.equal(nodePaths);
      parentProviderMock.verifyAll();
    });

  });

  describe("getNodeKey", () => {

    it("returns node key", () => {
      const key = createRandomECInstanceNodeKey();
      const treeNode = createRandomTreeNodeItem(key);

      parentProviderMock.setup((x) => x.getNodeKey(treeNode)).returns(() => key);
      const result = provider.getNodeKey(treeNode);
      expect(result).to.deep.equal(key);
    });

  });

  const constantFilter = "test";
  const filteredNodePaths: NodePathElement[] = [];

  filteredNodePaths[0] = createRandomNodePathElement();
  filteredNodePaths[0].node.label = "A-1";
  filteredNodePaths[0].filteringData = { matchesCount: 0, childMatchesCount: 1 };

  filteredNodePaths[0].children = [];
  filteredNodePaths[0].children[0] = createRandomNodePathElement();
  filteredNodePaths[0].children[0].node.label = "A-1-1 test";
  filteredNodePaths[0].children[0].filteringData = { matchesCount: 1, childMatchesCount: 0 };

  filteredNodePaths[1] = createRandomNodePathElement();
  filteredNodePaths[1].node.label = "A-2 test";
  filteredNodePaths[1].filteringData = { matchesCount: 1, childMatchesCount: 0 };

  filteredNodePaths[1].children = [];
  filteredNodePaths[1].children[0] = createRandomNodePathElement();
  filteredNodePaths[1].children[0].node.label = "A-2-1";
  filteredNodePaths[1].children[0].filteringData = { matchesCount: 0, childMatchesCount: 0 };

  filteredNodePaths[1].children[1] = createRandomNodePathElement();
  filteredNodePaths[1].children[1].node.label = "A-2-2";
  filteredNodePaths[1].children[1].filteringData = { matchesCount: 0, childMatchesCount: 0 };

  describe("countFilteringResults", () => {
    it("all matches get counted", () => {
      expect(provider.countFilteringResults(filteredNodePaths)).to.be.eq(2);
    });

    it("doesn't count if node paths don't have filtering data", () => {
      const paths: NodePathElement[] = [];
      paths[0] = createRandomNodePathElement();
      paths[0].node.label = "A-1";
      paths[0].filteringData = undefined;
      expect(provider.countFilteringResults(paths)).to.eq(0);
    });
  });

  describe("getActiveMatch", () => {
    it("returns correct match", () => {
      provider = new FilteredPresentationTreeDataProvider(parentProviderMock.object, constantFilter, filteredNodePaths);
      const result = provider.getActiveMatch(2);

      expect(result).to.not.be.undefined;
      expect(result!.nodeId).to.be.eq(createTreeNodeItem(filteredNodePaths[1].node).id);
      expect(result!.matchIndex).to.be.eq(0);
    });

    it("returns undefined when index is 0 or lower", () => {
      provider = new FilteredPresentationTreeDataProvider(parentProviderMock.object, constantFilter, filteredNodePaths);
      const result = provider.getActiveMatch(0);
      expect(result).to.be.undefined;
    });
  });
});
