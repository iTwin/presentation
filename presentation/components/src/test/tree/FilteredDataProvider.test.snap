// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`FilteredTreeDataProvider getNodes returns child nodes 1`] = `
Array [
  Object {
    "__presentation-components/key": Object {
      "instanceKey": Object {
        "className": "Auto Loan Account",
        "id": "0xb1d8000000de35",
      },
      "pathFromRoot": Array [
        "66dcb686-2502-41e0-a1c3-9b08a0214f59",
        "43142b23-f95f-40db-b473-9fde2c0a107d",
      ],
      "type": "ECInstanceNode",
    },
    "description": "Doloribus aut culpa ratione vel molestias omnis.",
    "icon": "reboot",
    "id": "43142b23-f95f-40db-b473-9fde2c0a107d/66dcb686-2502-41e0-a1c3-9b08a0214f59",
    "label": "A-2-1",
    "labelDefinition": PropertyRecord {
      "property": Object {
        "displayLabel": "Label",
        "name": "node_label",
        "typename": "string",
      },
      "value": Object {
        "displayValue": "Realigned",
        "value": "bluetooth",
        "valueFormat": 0,
      },
    },
    "parentId": "680c80a2-8cf0-49ee-8272-ec6682ae559c/e412801a-e3e4-4479-8999-2f94b3186fb7",
    "style": Object {
      "colorOverrides": Object {
        "backgroundColor": 1126240511,
      },
    },
  },
  Object {
    "__presentation-components/key": Object {
      "instanceKey": Object {
        "className": "Village",
        "id": "0x143170000017783",
      },
      "pathFromRoot": Array [
        "23ea98de-9451-4bb8-8be2-ea063b9cf6fe",
        "83e0ae64-7290-487e-8e56-9f025212b800",
      ],
      "type": "ECInstanceNode",
    },
    "autoExpand": true,
    "checkBoxState": 1,
    "description": "Rem quia odit.",
    "hasChildren": true,
    "icon": "Avon",
    "id": "83e0ae64-7290-487e-8e56-9f025212b800/23ea98de-9451-4bb8-8be2-ea063b9cf6fe",
    "isCheckboxVisible": true,
    "label": "A-2-2",
    "labelDefinition": PropertyRecord {
      "property": Object {
        "displayLabel": "Label",
        "name": "node_label",
        "typename": "string",
      },
      "value": Object {
        "displayValue": "end-to-end",
        "value": "Licensed",
        "valueFormat": 0,
      },
    },
    "parentId": "680c80a2-8cf0-49ee-8272-ec6682ae559c/e412801a-e3e4-4479-8999-2f94b3186fb7",
  },
]
`;

exports[`FilteredTreeDataProvider getNodes returns root nodes 1`] = `
Array [
  Object {
    "__presentation-components/key": Object {
      "instanceKey": Object {
        "className": "Congo",
        "id": "0x71220000010e43",
      },
      "pathFromRoot": Array [
        "024f2316-265a-46e8-965a-837e308ae678",
        "62f3ccbf-c51b-48ed-ad1d-0420ea196af6",
      ],
      "type": "ECInstanceNode",
    },
    "autoExpand": true,
    "description": "Nobis repellendus fugiat velit.",
    "hasChildren": true,
    "icon": "Shoes",
    "id": "62f3ccbf-c51b-48ed-ad1d-0420ea196af6/024f2316-265a-46e8-965a-837e308ae678",
    "isCheckboxDisabled": true,
    "isCheckboxVisible": true,
    "label": "A-1",
    "labelDefinition": PropertyRecord {
      "property": Object {
        "displayLabel": "Label",
        "name": "node_label",
        "typename": "string",
      },
      "value": Object {
        "displayValue": "methodologies",
        "value": "capacitor",
        "valueFormat": 0,
      },
    },
  },
  Object {
    "__presentation-components/key": Object {
      "instanceKey": Object {
        "className": "payment",
        "id": "0x1590a0000003157",
      },
      "pathFromRoot": Array [
        "e412801a-e3e4-4479-8999-2f94b3186fb7",
        "680c80a2-8cf0-49ee-8272-ec6682ae559c",
      ],
      "type": "ECInstanceNode",
    },
    "autoExpand": true,
    "hasChildren": true,
    "id": "680c80a2-8cf0-49ee-8272-ec6682ae559c/e412801a-e3e4-4479-8999-2f94b3186fb7",
    "isCheckboxVisible": true,
    "label": "A-2",
    "labelDefinition": PropertyRecord {
      "property": Object {
        "displayLabel": "Label",
        "name": "node_label",
        "typename": "string",
      },
      "value": Object {
        "displayValue": "Union",
        "value": "Gorgeous",
        "valueFormat": 0,
      },
    },
  },
]
`;
